### ISSUE TITLE
Question on the Value of Training Loss for DiffuLoss with MAR and Causal Methods

### ISSUE THREAD

**Issue by bugWholesaler** [LTH14/mar#20]
Thanks for your great work ！
I am currently engaged in a project that involves the DiffuLoss, and I am curious about the convergence behavior of the training loss. Specifically, I would like to know how much the training loss can ultimately converge for both the MAR and Causal methods. And which converges faster, the MAR method or the Causal method?
Again，awesome work！I look forward to your reponse！

**Comment by LTH14** [comment]
Thanks for your interest! Here is the 800 epochs training loss curve on ImageNet, MAR-L: 
<img width="985" alt="image" src="https://github.com/user-attachments/assets/22a2b14b-eba1-4612-8def-beba999b85da">
Similar to DiT, our training loss will never really converge: training for longer will keep improving the performance. However, at 400 epochs the performance (FID) is typically already quite good (< 2). In our experience, the MAR method achieves much better performance than the Causal method on ImageNet.

**Comment by yuhuUSTC** [comment]
Thanks for the great work! 
I got similar loss convergence curve when training the model myself. I find that the averaged diffloss quickly converges to about 0.25 and stops decreasing further. However, the generation performance continually becomes better as training going. This confuses me. On one side, the training loss seems to stop converging but the generation is increasingly better. On the other side, from my understanding, the averaged 0.25 Diffloss is still pretty large given the l2 loss implementation in diffloss. This means that the averaged per pixel distance to the Gaussion noise is 0.5, which is far from optimal. 
I look forward to your reponse！

**Comment by LTH14** [comment]
Thanks for your interest. This repo should achieve a loss curve similar to the above one (around 0.33), but different data could result in slightly different loss values. This kind of quickly converged loss curve is commonly observed for diffusion loss (e.g. Figure 13 in [DiT](https://arxiv.org/pdf/2212.09748)). However, after the initial quick decrease, it still constantly decreases as shown in the figure above, and thus keeps improving the generation performance.

The absolute loss value is related to many factors: tokenizer, dataset, model capacity, noise scheduling, etc. The absolute value can thus vary a lot. For example, DiT's loss is around 0.15. This is because the denoising function is very hard to learn, especially when the noise level is high. Moreover, since we use a very large masking ratio (randomly sampled between 0.7 and 1.0) during MAR's training, our loss is even larger than DiT.

**Comment by yuhuUSTC** [comment]
Thanks for the answer.

**Comment by Robootx** [comment]
how about the loss of causal method

**Comment by LTH14** [comment]
@Robootx here is the loss for random order causal method (with teacher-forcing language modeling loss): 
<img width="499" alt="image" src="https://github.com/user-attachments/assets/6ac82db6-e462-4f8c-9c14-8e3906d41a77">

**Comment by Robootx** [comment]
> @Robootx here is the loss for random order causal method (with teacher-forcing language modeling loss): <img alt="image" width="499" src="https://private-user-images.githubusercontent.com/22166952/361051338-6ac82db6-e462-4f8c-9c14-8e3906d41a77.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjQ0ODIyNTQsIm5iZiI6MTcyNDQ4MTk1NCwicGF0aCI6Ii8yMjE2Njk1Mi8zNjEwNTEzMzgtNmFjODJkYjYtZTQ2Mi00ZjhjLTljMTQtOGUzOTA2ZDQxYTc3LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA4MjQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwODI0VDA2NDU1NFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWYxYWNlNWQxNmQzZmViNDdjNzQxNGRiZjY2ZjU5MTNjMjNjYTkwMmZhMzZiOGY4MjY1NWQwOTQxODhjOGJmZGYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.aWbDDZhOKWQB4GetLVeVvIyQPOjIiv9BxLwMAOtXd0s">

thank you very much

**Comment by Robootx** [comment]
Could you please show me some images generated by a causal method?

**Comment by zythenoob** [comment]
I wonder how the model performs at different training stages, e.g., how many training steps it takes to be able to generate the shape of an object?

**Comment by Juhywcy** [comment]
> Thanks for your interest! Here is the 800 epochs training loss curve on ImageNet, MAR-L: <img alt="image" width="985" src="https://private-user-images.githubusercontent.com/22166952/358216994-22a2b14b-eba1-4612-8def-beba999b85da.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjUyNzc4MTIsIm5iZiI6MTcyNTI3NzUxMiwicGF0aCI6Ii8yMjE2Njk1Mi8zNTgyMTY5OTQtMjJhMmIxNGItZWJhMS00NjEyLThkZWYtYmViYTk5OWI4NWRhLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA5MDIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwOTAyVDExNDUxMlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTAwY2IzZWY3MzA3ODYyNmYxMGIyNjNjNDM0NWQwZjYyZGRlYmI2MzUxNGI1NTllZWUxYjk2OTlhMmUyYzQ1NGUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.6duqa5_6YbTcG29fbr_FHZ4R5B7qAeGvscfsPu9Ux68"> Similar to DiT, our training loss will never really converge: training for longer will keep improving the performance. However, at 400 epochs the performance (FID) is typically already quite good (< 2). In our experience, the MAR method achieves much better performance than the Causal method on ImageNet.

Thanks for your loss curve.
Can you give me the loss curve of Mar_huge, Thanks!

**Comment by LTH14** [comment]
@Juhywcy 
<img width="981" alt="image" src="https://github.com/user-attachments/assets/987c17d5-1cca-4fbb-8d54-594287a94504">

**Comment by Juhywcy** [comment]
> @Juhywcy <img alt="image" width="981" src="https://private-user-images.githubusercontent.com/22166952/363693613-987c17d5-1cca-4fbb-8d54-594287a94504.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjUyOTA2NTEsIm5iZiI6MTcyNTI5MDM1MSwicGF0aCI6Ii8yMjE2Njk1Mi8zNjM2OTM2MTMtOTg3YzE3ZDUtMWNjYS00ZmJiLThkNTQtNTk0Mjg3YTk0NTA0LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA5MDIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwOTAyVDE1MTkxMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTczMzczYjFiODIyMjdiOGMzYTZlM2ViZGFkNTYyMTFlZmViOGZiNDJiODRlZTg0MTdjNWUzNjYxNWFhNjdjNGImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.pwPxna1m0z-peFm-wrHKD_04gUFd13wmMW29caCx-fU">

thanks!

**Comment by Juhywcy** [comment]
can you prepare the lr details of mar_huge training?  It will help reproduce the results.

**Comment by LTH14** [comment]
@Juhywcy the learning rate schedule and value is the same for all models

**Comment by Juhywcy** [comment]
> @Juhywcy the learning rate schedule and value is the same for all models

thanks for your reply!! constant and 1e-4?

**Comment by Juhywcy** [comment]
> > > @Juhywcy the learning rate schedule and value is the same for all models
> > 
> > 
> > thanks for your reply!! constant and 1e-4?
> 
> yes -- also with linear warmup.

thanks for your fast reply! have a good day!

**Comment by poppuppy** [comment]
Thanks for your great work!
Can you also provide the loss when training MAR with cross entropy loss? Thank you, and look forward to your reply.

**Comment by LingweiMeng** [comment]
> > > @Juhywcy the learning rate schedule and value is the same for all models
> > 
> > 
> > thanks for your reply!! constant and 1e-4?
> 
> yes -- also with linear warmup.

I just want to confirm that, in the paper, the lr is 8e-4?
Why it is 1e-4 here?
Thank you.  :)

**Comment by LTH14** [comment]
@LingweiMeng we scale the final learning rate according to the total batch size divided by 256. 1e-4 is the "base learning rate" before scaling.

**Comment by LingweiMeng** [comment]
Thank you.
